cli/pull.go

```diff
@@ -61,36 +63,62 @@ func (p *PullCommand) runPull(args []string) error {
 	}
 	defer responseBody.Close()
 
-	return renderOutput(responseBody)
+	return showProgress(responseBody)
 }
 
-// renderOutput draws the commandline output via api response.
-func renderOutput(responseBody io.ReadCloser) error {
+// bufwriter defines interface which has Write and Flush behaviors.
+type bufwriter interface {
+	Write([]byte) (int, error)
+	Flush() error
+}
+
+// showProgress shows pull progress status.
+func showProgress(body io.ReadCloser) error {
 	var (
-		start = time.Now()
-		fw    = progress.NewWriter(os.Stdout)
+		output bufwriter = bufio.NewWriter(os.Stdout)
+
+		start      = time.Now()
+		isTerminal = terminal.IsTerminal(int(os.Stdout.Fd()))
 	)
 
-	dec := json.NewDecoder(responseBody)
+	if isTerminal {
+		output = progress.NewWriter(os.Stdout)
+	}
+
+	dec := json.NewDecoder(body)
 	if _, err := dec.Token(); err != nil {
 		return fmt.Errorf("failed to read the opening token: %v", err)
 	}
 
+	refStatus := make(map[string]string)
 	for dec.More() {
-		var objs []ctrd.ProgressInfo
-
-		tw := tabwriter.NewWriter(fw, 1, 8, 1, ' ', 0)
+		var infos []ctrd.ProgressInfo
 
-		if err := dec.Decode(&objs); err != nil {
+		if err := dec.Decode(&infos); err != nil {
 			return fmt.Errorf("failed to decode: %v", err)
 		}
 
-		if err := display(tw, objs, start); err != nil {
-			return err
+		// only display the new status if the stdout is not terminal
+		if !isTerminal {
+			newInfos := make([]ctrd.ProgressInfo, 0)
+			for i, info := range infos {
+				old, ok := refStatus[info.Ref]
+				if !ok || (ok && info.Status != old) {
```

Comment:
We can simplify this with `if  !ok || info.Status != old `

Reply:
Make senses. Updated